src/goto-instrument/replace_function_bodies.cpp

```diff
@@ -0,0 +1,388 @@
+  /// @param function whose body to generate
+  /// @param symbol_table of the current goto program
+  /// @param function_name Identifier of function
+  virtual void generate_function_body(
+    goto_functiont &function,
+    symbol_tablet &symbol_table,
+    irep_idt function_name) = 0;
+
+public:
+  virtual ~replace_function_bodiest() = default;
+
+  /// Replace the function body with one based on the implementation
+  /// This will work the same whether or not the function already has a body
+  /// @param function whose body to replace
+  /// @param symbol_table of the current goto program
+  /// @param function_name Identifier of function
+  void replace_function_body(
+    goto_functiont &function,
+    symbol_tablet &symbol_table,
+    irep_idt function_name)
+  {
+    function.body.clear();
+    generate_parameter_names(function, symbol_table, function_name);
+    generate_function_body(function, symbol_table, function_name);
+  }
+
+private:
+  /// Generate parameter names for unnamed parameters.
+  /// CBMC expect functions to have parameter names
+  /// if they are called and have a body
+  void generate_parameter_names(
+    goto_functiont &function,
+    symbol_tablet &symbol_table,
+    irep_idt function_name) const
+  {
+    namespacet ns(symbol_table);
+    int param_counter = 0;
+    for(auto &parameter : function.type.parameters())
+    {
+      if(parameter.get_identifier().empty())
+      {
+        const std::string param_base_name =
+          parameter.get_base_name().empty()
+            ? "__param$" + std::to_string(param_counter++)
+            : id2string(parameter.get_base_name());
+        irep_idt new_param_identifier =
+          id2string(function_name) + "::" + param_base_name;
+        parameter.set_base_name(param_base_name);
+        parameter.set_identifier(new_param_identifier);
+        parameter_symbolt new_param_sym;
+        new_param_sym.name = new_param_identifier;
+        new_param_sym.type = parameter.type();
+        new_param_sym.base_name = param_base_name;
+        new_param_sym.mode = ID_C;
+        auto const &function_symbol = symbol_table.lookup_ref(function_name);
+        new_param_sym.module = function_symbol.module;
+        new_param_sym.location = function_symbol.location;
+        symbol_table.add(new_param_sym);
+      }
+    }
+    auto &function_symbol = symbol_table.get_writeable_ref(function_name);
+    function_symbol.type = function.type;
+  }
+};
+
+class assume_false_replace_function_bodiest : public replace_function_bodiest
+{
+protected:
+  void generate_function_body(
+    goto_functiont &function,
+    symbol_tablet &symbol_table,
+    irep_idt function_name) override
+  {
+    auto assume_instruction = function.body.add_instruction();
+    assume_instruction->make_assumption(false_exprt());
+    assume_instruction->code = code_assumet(false_exprt());
+    auto end_instruction = function.body.add_instruction();
+    end_instruction->make_end_function();
```

Comment:
Set up the `source_location` (to the location of the function symbol) and the `function` (to `function_name`) members of each instruction.